### ISSUE TITLE
model definition changes between training and testing

### ISSUE THREAD

**Issue by FraPochetti** [ultralytics/yolov5#2780]
## ❔Question
Hi, 
I noticed in `train.py` you define the model instantiating `Model` ([here](https://github.com/ultralytics/yolov5/blob/1487bc84ff3babfb502dffb5ffbdc7e02fcb1879/train.py#L88)), whereas in `detect.py` you make use of `attempt_load` ([here](https://github.com/ultralytics/yolov5/blob/1487bc84ff3babfb502dffb5ffbdc7e02fcb1879/detect.py#L34)) which invokes an ensemble strategy in the backend.
Can you share any insights around why you are doing this?

What about using `Model` (as done in `train.py`) with a sepcific config (small, medium, large architecture) and then setting the model to `eval` mode?
This is what we are doing in [icevision](https://github.com/airctic/icevision). Anything wrong with this approach?

Thanks Glenn!

**Comment by glenn-jocher** [comment]
👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. 

**train.py can either build a model based on a yaml architecture config file or load a pretrained model. test.py simply loads pretrained models.**

A few important aspects of each:

## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): 0.001
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): 0.6
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): True
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): None

## test.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): 0.001
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): 0.6
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): True
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): 0.5 * maximum stride

## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): 0.25 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): 0.45
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): False
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): None

## YOLOv5 PyTorch Hub

- autoShape() class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

**Comment by FraPochetti** [comment]
@glenn-jocher Thanks so much as always for the detailed response.
Regardless of the fact that you are optimizing data+models for specific use cases, there is nothing wrong in defining a `Model` object (using either a pretrained model or a YAML config) and put it in `eval` mode to run predictions, right?
Maybe I won't get the best possible results as I am not ensembling, but is it still ok?

**Comment by glenn-jocher** [comment]
@FraPochetti ensembling is natively supported by the command line functions like detect.py and test.py but of course is not required for inference.

There's no optimization really, train.py is required to build models, and test.py, detect.py are required to load models, so they use different methods, that's all.

Both methods should produce the same model in the end if all layers are transferred.

**Comment by FraPochetti** [comment]
Thanks!
